### DISCUSSION TITLE
Error: EISDIR: illegal operation on a directory, read

### DISCUSSION THREAD

**Discussion by craigbloodworth** [vercel/vercel#4711] [Help] [answered]
Every time a source file is changed, the rebuild is triggered and the browser is refreshed the my vercel instance fails with the error:
```
Unhandled rejection: [Error: EISDIR: illegal operation on a directory, read] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'read'
}
LambdaError: RequestId: 2e33f94b-d417-4c42-a214-081a06533adf Process exited before completing request
    at Lambda.<anonymous> (/usr/local/lib/node_modules/vercel/dist/index.js:2:2755509)
    at Generator.next (<anonymous>)
    at fulfilled (/usr/local/lib/node_modules/vercel/dist/index.js:2:2753396)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```
The only way to continue is to reset the cli and run it again. Has anybody seen this before & any hints in how I could fix it?

**Comment by rylandking18** [comment]
Pretty sure I'm getting an error that's related... 

![Screen Shot 2020-06-24 at 7 04 17 AM](https://user-images.githubusercontent.com/38166329/85561578-f2df9300-b5e8-11ea-8c3d-79df5ab5524f.jpg)

**Reply by craigbloodworth** [reply]
Hmmm interesting. I should have mentioned this is running in dev mode using the CLI but interesting to see it's happening in the hosted environment too

**Comment by sdwvit** [comment]
Have this very same error happening in production build. Pulling my hairs rn. 
![image](https://user-images.githubusercontent.com/8777372/109346490-c1862980-783f-11eb-9820-28515d1e6f7e.png)

**Reply by sdwvit** [reply]
ok the issue in my case was happening after the build before uploading the artifacts.
My package.json was containing a dependency to the local file, higher one level than the package.json. Changing that dependency to point to npmjs instead solved an issue for me.
I think Vercel changed something and stopped including files outside of the build folder, causing the deployment to fail.

I created a ticket to vercel to help me debug this thing, let's hear from them too.

**Reply by atilafassina** [reply]
on your project settings where you set the root of your project deploy you have a checkbox underneath where you should be able to define including files outside of the root
![Screenshot 2021-02-27 at 14 33 23](https://user-images.githubusercontent.com/2382552/109388682-e5fa0880-7908-11eb-9923-5b636ddda737.png)

**Reply by sdwvit** [reply]
@atilafassina it's checked

**Reply by sdwvit** [reply]
same build succeeds with yarn commands and fails with npm commands

**Comment by saimeunt** [comment]
As far as my setup is concerned, I had a directory named .env at the root of my Next.js project, containing my environment variables for development/staging/production.
I would then use dotenv to load the correct set of variables in production vs development.

Problem is, Next.js standard environment variables setup allows to specify a .env (AS A FILE) at the root of your project to specify default environment variables for all environments. (https://nextjs.org/docs/basic-features/environment-variables#default-environment-variables).

So basically the vercel CLI was failing with EISDIR because it tried to load my .env folder as .env file, very hard to debug!

**Comment by bryan-matheus** [comment]
I also had this same problem. In my case I was using the `pnpm` package manager. I managed to fix it like:

1. Remove the `node_modules` folder
2. Removed the `pnpm-lock.yaml` file
3. Ran `yarn install`

**Comment by ycjcl868** [comment]
Having the same problem, I've updated `pnpm-lock.yaml`.

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/13595509/159196400-993225c9-1e33-4d20-9c7c-c8e9e26bb13a.png">

**Reply by nicksrandall** [reply]
I'm also using Remix + PNPM and I have this exact same issue.

**Reply by mauricekleine** [reply]
Same here. When I switch to `npm`, this issue does not happen and the build succeeds. But when I switch back to using `pnpm`, I get the same error again.

**Reply by vasco3** [reply]
Have this issue with Remix + PNPM. If I click on the *Redeploy* button it works. But when I visit the app it throws a 404

**Reply by vasco3** [reply]
@ycjcl868 did you resolve this?

**Reply by ycjcl868** [reply]
> @ycjcl868 did you resolve this?

There is no solution yet.

**Reply by loucyx** [reply]
Kinda weird behavior because vercel's deploy shoudln't care about the package manager we use locally. Hope this gets resolved, in the meantime I guess I'll just use `npm` for my remix project -_-

**Comment by corwin-of-amber** [comment]
I was having a similar issue for a pretty long time whenever my directory contained symlinks. I am using a script on `vercel-build` that removes symlinks and replaces them with a copy of the linked directory.

**Reply by Gomah** [reply]
@corwin-of-amber Do you have a code snippet to share? We're facing a similar issue

**Comment by QGB** [comment]
Cloning github.com/qgbcs/vercel-django-example (Branch: main, Commit: 80b1f7e)
Cloning completed: 631.319ms
Warning: Due to `builds` existing in your configuration file, the Build and Development Settings defined in your Project Settings will not apply. Learn More: https://vercel.link/unused-build-settings
Installing build runtime...
Build runtime installed: 3.195s
Looking up build cache...
Build Cache not found
Installing required dependencies...
[Error: EISDIR: illegal operation on a directory, open '/vercel/path0/vc__handler__python.py'] {
  errno: -21,
  code: 'EISDIR',
  syscall: 'open',
  path: '/vercel/path0/vc__handler__python.py'
}

**Comment by ianmat55** [comment]
I know this is very old, but the way I solved my particular run-in with this bug was by removing `next export` from my preset build script
`"build": "next build && next export"` --> `"build": "next build",`. Something to do with image optimization using the default loader not being compatible with next export

**Comment by swarnava** [comment, marked as answer]
We have pushed a fix for this today. If you are still facing any issue, please create a ticket here: https://vercel.com/help
